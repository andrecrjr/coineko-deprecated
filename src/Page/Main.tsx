import React from 'react';
import { Outlet } from 'react-router';
import Header from 'src/Components/Header';

export const Main = () => {
	return (
		<>
			<div className="App">
				<Header />
				<section className="flex flex-col justify-center sm:items-center ml-2 sm:ml-0 relative">
					<Outlet />
				</section>
			</div>
		</>
	);
};
