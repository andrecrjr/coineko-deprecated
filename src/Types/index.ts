import { Dispatch } from 'react';
import currencyList from './currencyListApi.json';

export type PageCurrencyQuery = {
	vs_currency: string;
	order: string;
	per_page: string;
	page: string;
	sparkline: string;
	price_change_percentage: string;
	category?: string;
};

export type Currency = typeof currencyList;

export type CurrencyList = Currency[];

export type PaginationState = {
	number: number;
	firstOrPagination: 'current' | 'pagination';
};

export type Pagination = {
	page: PaginationState;
	setPagination: React.Dispatch<React.SetStateAction<PaginationState>>;
};

export type PortfolioType = {
	userCurrency: string[] | null;
	setPortfolio: Dispatch<{ type: string; payload: string }>;
};
