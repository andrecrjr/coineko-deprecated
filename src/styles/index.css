@import url("https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap");
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {

	body {
		@apply  bg-[#ECECEC] min-h-screen;
	}
	.table--head {
		@apply text-dark-purple-neko font-bold text-[0.77rem] py-[0.77rem] font-roboto;
		@apply sm:text-[0.77rem];
	}

	.table--body {
		@apply text-center text-[0.90rem] font-roboto font-bold pt-4;
		@apply lg:text-left;
	}
	.table--body:last-child {
		@apply pr-5;
	}

	.table--body__coin {
		@apply text-left text-[0.70rem] text-[#272727] leading-5 w-[180px];
		@apply lg:pl-0 sm:w-auto;
	}

	.table--body__border{
		@apply border-b-[24px] border-[transparent];
	}

	.table--body__line:last-child {
		@apply border-b-[24px] border-[transparent];
	}
	.search--box{
		@apply border-2;
	}
	.suggest--box{
		@apply absolute bg-white w-full z-10;
	}
}
