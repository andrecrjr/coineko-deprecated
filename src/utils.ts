import { PageCurrencyQuery } from './Types';

const styleCurrency = {
	style: 'currency',
	currency: 'USD'
};

export const formatterMoney = (
	language = 'en-US',
	currency: typeof styleCurrency,
	amount: number
) => {
	return new Intl.NumberFormat(language, currency).format(amount);
};

export const convertFilterQueryString = (
	filterPaginationAndCategory: PageCurrencyQuery
) => {
	if (typeof filterPaginationAndCategory?.category === 'undefined') {
		delete filterPaginationAndCategory['category'];
	}
	const filterResult = new URLSearchParams(
		filterPaginationAndCategory
	).toString();
	return filterResult;
};

export const storageObject = {
	get: <T>(storageKey: string): T => {
		const data = localStorage.getItem(storageKey) || JSON.stringify([]);
		return JSON.parse(data);
	},
	set: (storageKey: string, setData: string[]): boolean => {
		if (!setData) throw new Error('No data to set in localstorage');
		localStorage.setItem(storageKey, JSON.stringify(setData));
		return true;
	}
};

export const addToPortfolio = (newData: string | null) => {
	const portfolioData = storageObject.get<[]>('portfolio');
	if (newData) {
		return [...portfolioData, newData];
	}
	return [...portfolioData];
};
