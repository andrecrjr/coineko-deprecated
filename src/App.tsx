import Header from './Components/Header';
import { useReducer, useState } from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Routes, Route } from 'react-router-dom';
import {
	FilterGlobalContext,
	PaginationContext,
	PortfolioContext
} from './state/Contexts';
import { PaginationState } from './Types';
import TablePage from './Page/TablePage';
import { PortfolioReducer } from './state/Reducers/portfolio';
import { storageObject } from './utils';

function App() {
	const [page, setPagination] = useState<PaginationState>({
		firstOrPagination: 'current',
		number: 1
	});

	const [portfolio, dispatchPortfolio] = useReducer(
		PortfolioReducer,
		storageObject.get<string[]>('portfolio')
	);

	return (
		<Router>
			<PortfolioContext.Provider
				value={{ userCurrency: portfolio, setPortfolio: dispatchPortfolio }}
			>
				<PaginationContext.Provider value={{ page, setPagination }}>
					<FilterGlobalContext.Provider
						value={{
							vs_currency: 'usd',
							order: 'market_cap_desc',
							per_page: '50',
							sparkline: 'false',
							page: '1',
							price_change_percentage: '1h,24h,7d'
						}}
					>
						<AppRoutes />
					</FilterGlobalContext.Provider>
				</PaginationContext.Provider>
			</PortfolioContext.Provider>
		</Router>
	);
}

export const AppRoutes = () => {
	return (
		<div className="App">
			<Header />
			<Routes>
				<Route
					path="/"
					element={
						<TablePage description="Price of the main cryptocurrencies by Market Capitalization." />
					}
				/>
				<Route
					path="nft"
					element={
						<TablePage
							description="Ranking of NFT price by Market Capitalization."
							category={'non-fungible-tokens-nft'}
						/>
					}
				/>
				<Route
					path="de-fi"
					element={
						<TablePage
							description="Ranking of De-Fi prices by Market Capitalization."
							category={'decentralized-finance-defi'}
						/>
					}
				/>
				<Route
					path="exchange"
					element={
						<TablePage
							description="Price of cryptocurrencies exchange by Market Capitalization."
							category={'exchange-based-tokens'}
						/>
					}
				/>
			</Routes>
		</div>
	);
};

export default App;
