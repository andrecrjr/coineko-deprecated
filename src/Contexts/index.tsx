import { createContext } from 'react';
import { PageCurrencyQuery, Pagination } from 'src/Types';

export const PaginationContext = createContext<Pagination>({
	page: {
		number: 1,
		firstOrPagination: 'current'
	},
	setPagination: () => {
		return {
			number: 1,
			firstOrPagination: 'current'
		};
	}
});

export const FilterGlobalContext = createContext<PageCurrencyQuery>({
	vs_currency: 'usd',
	order: 'market_cap_desc',
	per_page: '50',
	page: '1',
	sparkline: 'false',
	price_change_percentage: '1h,24h,7d',
	category: 'cryptocurrency'
});
