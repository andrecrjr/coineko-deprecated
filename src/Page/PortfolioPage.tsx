import React from 'react';
import Table from 'src/Components/Table';
import { CurrencyList } from 'src/Types';

import { useGetPortfolio } from 'src/state/Hooks/usePortfolio';
import { useFetch } from 'src/state/Hooks/useSWR';

export const PortfolioPage = () => {
	const { portfolioFilterQuery, portfolioItens } = useGetPortfolio();
	const { data, isLoading } = useFetch<CurrencyList>(
		portfolioFilterQuery,
		'get'
	);

	const NotLoaded = () => {
		if (!isLoading && !data) {
			return <p>Not found any cryptocurrency in your browser</p>;
		}
		return <></>;
	};

	return (
		<>
			<section className="flex items-center mt-8 w-10/12 mb-2 md:mb-5 md:mt-10 ">
				<h3 className="text-xs text-left items-start md:text-base">
					Your Portfolio
				</h3>
			</section>
			{!isLoading &&
				data &&
				data.length > 0 &&
				portfolioItens &&
				portfolioItens?.length > 0 && <Table data={data} />}
			<NotLoaded />
		</>
	);
};
