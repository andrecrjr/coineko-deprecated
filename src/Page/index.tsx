import { useContext } from 'react';
import Table from 'src/Components/Table';
import { FilterGlobalContext, PaginationContext } from 'src/Contexts';
import { useFilter } from 'src/Hooks/useFilter';
import { useFetch } from 'src/Hooks/useSWR';
import { CurrencyList } from 'src/Types';
import Cat from 'src/assets/cat.svg?component';

type Props = {
	description: string;
	category?: string;
};

const TablePage = ({ description, category }: Props) => {
	const filterContext = useContext(FilterGlobalContext);

	const { page } = useContext(PaginationContext);
	const filterFetchData = useFilter({
		filterDataObject: {
			...filterContext,
			...{ page: page.number.toString() || '1', category }
		}
	});

	const { data: cryptoCurrenciesList, isLoading } = useFetch<CurrencyList>(
		filterFetchData,
		'GET'
	);

	return (
		<>
			{!isLoading && (
				<section className="flex flex-col justify-center sm:items-center ml-2 sm:ml-0 relative">
					<section className="flex items-center mt-8 w-10/12 mb-2 md:mb-5 md:mt-10 ">
						<span>
							<Cat className="w-[35px] h-[35px]" />
						</span>
						<h3 className="text-xs text-left items-start md:text-base">
							{description}
						</h3>
					</section>
					<Table data={cryptoCurrenciesList} />
				</section>
			)}
		</>
	);
};

export default TablePage;
