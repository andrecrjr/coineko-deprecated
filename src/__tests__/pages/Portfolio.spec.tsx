import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import { describe, expect, vi, it } from 'vitest';
import { ContainerWrapper } from './container';
import { PortfolioPage } from 'src/Page/PortfolioPage';
import portfolioMock from 'src/__mocks__/portfolio.mock.json';
import portfolioRemovedMock from 'src/__mocks__/portfolioremoved.mock.json';
import { storageObject } from 'src/utils';

vi.mock('axios', async () => {
	const create = vi.fn().mockImplementation(({ baseURL }) => {
		return {
			get: vi.fn().mockImplementation((url) => {
				const apiFilterUrl: string = baseURL + url;
				if (
					apiFilterUrl.includes(
						'https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=ethereum%2Cbitcoin%2Cbanano&order=market_cap_desc&per_page=100&page=1&sparkline=false'
					)
				) {
					return { data: portfolioMock };
				}
				if (
					apiFilterUrl.includes(
						'https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=ethereum%2Cbanano&order=market_cap_desc&per_page=100&page=1&sparkline=false'
					)
				) {
					return { data: portfolioRemovedMock };
				}
			})
		};
	});

	return { default: { create } };
});

describe('Portfolio page', () => {
	it('should render portfolio page title', async () => {
		render(
			<ContainerWrapper>
				<PortfolioPage />
			</ContainerWrapper>
		);
		expect((await screen.findByText('Your Portfolio')).textContent).toBe(
			'Your Portfolio'
		);
	});

	it('should not render portfolio user table', async () => {
		vi.spyOn(storageObject, 'get').mockReturnValue([]);
		render(
			<ContainerWrapper>
				<PortfolioPage />
			</ContainerWrapper>
		);
		expect(
			(await screen.findByText('Not found any cryptocurrency in your browser'))
				.textContent
		).toBe('Not found any cryptocurrency in your browser');
	});

	it('should render portfolio user table', async () => {
		vi.spyOn(storageObject, 'get').mockReturnValue([
			'ethereum',
			'bitcoin',
			'banano'
		]);
		const { container } = render(
			<ContainerWrapper>
				<PortfolioPage />
			</ContainerWrapper>
		);

		expect((await screen.findByText('Bitcoin')).textContent).toBe('Bitcoin');
		expect((await screen.findByText('Ethereum')).textContent).toBe('Ethereum');
		expect((await screen.findByText('Banano')).textContent).toBe('Banano');
		expect(container.children).toMatchSnapshot();
	});

	it('should not render currency removed in portfolio', async () => {
		vi.spyOn(storageObject, 'get').mockReturnValue([
			'ethereum',
			'bitcoin',
			'banano'
		]);
		render(
			<ContainerWrapper>
				<PortfolioPage />
			</ContainerWrapper>
		);
		const bitcoinFavoriteButton = (
			await screen.findAllByTestId('favorite-crypto')
		)[0];
		fireEvent.click(bitcoinFavoriteButton);
		//hack to wait the screen load
		await waitFor(() => ({}));
		expect(screen.queryByText(/Bitcoin/i)).toBeNull();
		expect(screen.queryByText(/Banano/i)).not.toBeNull();
	});
});
