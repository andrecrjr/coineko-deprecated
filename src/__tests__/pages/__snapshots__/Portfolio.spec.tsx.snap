// Vitest Snapshot v1

exports[`Portfolio page > should render portfolio user table 1`] = `
<section
  class="flex items-center mt-8 w-10/12 mb-2 md:mb-5 md:mt-10 "
>
  <h3
    class="text-xs text-left items-start md:text-base"
  >
    Your Portfolio
  </h3>
</section>
`;
